src/gas-price/gas-price.ts

```diff
@@ -0,0 +1,236 @@
+ * @param chainId
+ * @param providerName
+ * @param sanitizationSamplingWindow
+ */
+export const clearExpiredStoreGasPrices = (
+  chainId: string,
+  providerName: string,
+  sanitizationSamplingWindow: number
+) => {
+  // Remove gasPrices older than the sanitizationSamplingWindow
+  gasPriceStore[chainId]![providerName]!.gasPrices = gasPriceStore[chainId]![providerName]!.gasPrices.filter(
+    (gasPrice) => gasPrice.timestamp >= Date.now() - sanitizationSamplingWindow * 60 * 1000
+  );
+};
+
+/**
+ * Saves last transaction details into the store.
+ * @param chainId
+ * @param providerName
+ * @param nonce
+ */
+export const setLastOnChainDatafeedValues = (
+  chainId: string,
+  providerName: string,
+  dataFeedId: string,
+  dataFeedValues: { value: ethers.BigNumber; timestamp: number }
+) => {
+  initializeGasStore(chainId, providerName);
+  gasPriceStore[chainId]![providerName]!.lastOnChainDataFeedValues[dataFeedId] = dataFeedValues;
+};
+
+/**
+ * Removes last transaction details from the store.
+ * @param chainId
+ * @param providerName
+ * @param nonce
+ */
+export const clearLastOnChainDatafeedValue = (chainId: string, providerName: string, dataFeedId: string) => {
+  if (gasPriceStore[chainId]![providerName]!.lastOnChainDataFeedValues[dataFeedId]) {
+    delete gasPriceStore[chainId]![providerName]!.lastOnChainDataFeedValues[dataFeedId];
+  }
+};
+
+// TODO: Copied from airnode-utilities, should we import instead?
+export const getPercentile = (percentile: number, array: ethers.BigNumber[]) => {
+  if (array.length === 0) return;
+
+  array.sort((a, b) => (a.gt(b) ? 1 : -1));
+  const index = Math.ceil(array.length * (percentile / 100)) - 1;
+  return array[index];
+};
+
+export const multiplyGasPrice = (gasPrice: ethers.BigNumber, gasPriceMultiplier: number) =>
+  gasPrice.mul(ethers.BigNumber.from(Math.round(gasPriceMultiplier * 100))).div(ethers.BigNumber.from(100));
+
+/**
+ * Calculates the multiplier to use for pending transactions.
+ * @param recommendedGasPriceMultiplier
+ * @param scalingMultiplier
+ * @param lag
+ * @param scalingWindow
+ * @returns
+ */
+export const calculateScalingMultiplier = (
+  recommendedGasPriceMultiplier: number,
+  scalingMultiplier: number,
+  lag: number,
+  scalingWindow: number
+) => recommendedGasPriceMultiplier + (scalingMultiplier - recommendedGasPriceMultiplier) * (lag / scalingWindow);
+
+/**
+ * Fetches the provider recommended gas price and saves it in the store.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @returns {ethers.BigNumber}
+ */
+export const updateGasPriceStore = async (chainId: string, providerName: string, rpcUrl: string) => {
+  const provider = new ethers.providers.StaticJsonRpcProvider(rpcUrl, {
+    chainId: Number.parseInt(chainId, 10),
+    name: chainId,
+  });
+
+  // Get the provider recommended gas price
+  const goGasPrice = await go(async () => provider.getGasPrice(), { retries: 1, attemptTimeoutMs: 2000 });
```

Comment:
Not sure how I feel about retrying the get gas price call. It seems a bit arbitrary to me.

Reply:
Errors should probably be handled at the `getAirseekerV2ProviderRecommendedGasPrice` level in the main loop, with no `go` around the `getGasPrice` if we don't want to retry there.